sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs

```diff
@@ -1225,62 +1225,174 @@ private async Task RunProcessingAsync(CancellationToken cancellationToken)
+                                                                 ProcessingStoppedReason reason,
+                                                                 CancellationToken cancellationToken)
         {
-            // TODO:  Implement this;  current code is just a place holder.
-            if (ActivePartitionProcessors.TryRemove(partitionId, out var processor))
+            cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+            Logger.EventProcessorPartitionProcessingStop(partitionId, Identifier, EventHubName, ConsumerGroup);
+
+            var partition = default(TPartition);
+
+            try
             {
-                processor.CancellationSource.Cancel();
+                // If the partition processor is not being tracked or could not be retrieved from the tracking items,
+                // then it cannot be stopped.
+
+                if (!ActivePartitionProcessors.TryRemove(partitionId, out var partitionProcessor))
+                {
+                    return false;
+                }
+
+                // Attempt to stop the processor; any exceptions should be treated as a problem with processing, not
+                // associated with the attempt to stop.
+
+                partition = partitionProcessor.Partition;
 
                 try
                 {
-                    await processor.ProcessingTask.ConfigureAwait(false);
+                    partitionProcessor.CancellationSource.Cancel();
+                    await partitionProcessor.ProcessingTask.ConfigureAwait(false);
                 }
                 catch (TaskCanceledException)
                 {
-                    // nop
+                    // This is expected; no action is needed.
                 }
-                finally
+                catch
                 {
-                    processor.CancellationSource.Dispose();
+                    // The processing task is in a failed state; any logging and dispatching
+                    // to the error handler happened in the processing task before the exception
+                    // was thrown.  All that remains is to override the reason for stopping.
+
+                    reason = ProcessingStoppedReason.OwnershipLost;
                 }
+
+                // Clean up the partition processing resources that need to be disposed.
+
+                partitionProcessor.CancellationSource.Dispose();
+                partitionProcessor.ProcessingTask.Dispose();
+
+                // Notify the handler of the now-closed partition in a fire-and-forget manner.  The processor does not
+                // assume responsibility for observing or surfacing exceptions that may occur in the handler, so the associated
+                // task is intentionally unobserved.
+
+                cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+                _ = Task.Run(() => OnPartitionProcessingStoppedAsync(partition, reason, cancellationToken));
```

Comment:
Should the be `CancellationToken.None` similar to what we do in InvokeOnProcessingErrorAsync where we launch a fire and forget task?

Reply:
Thinking about this a little more - by not awaiting here we're also allowing calls to `OnPartitionProcessingStoppedAsync` to interleave with all the other "lifecycle" calls.  Is that what we actually want?  Feels like a difficult model to program in.

What do you think about awaiting here, and possibly calling the ErrorHandler (in a fire and forget fashion) if the call faults?

Reply:
Yeah; as a background task, it definitely should be `CancellationToken.None`, that's my bad.

I struggled a bit on what to do here; the contract that we have cross-language is that we explicitly do not surface exceptions from developer-provided code. 
 We do deviate from that during initialization, because we depend on the output of that call  - so there is precedent.  In a nutshell I see our options as:

- Follow the "your code needs to manage its own exceptions and we make no guarantees on behavior" model and run this in the background as is done.  This allows the processor to continue processing with maximum throughput, as we don't on handler calls that we don't need the result of.   As you mentioned, that does potentially interleave calls is a partition is claimed and surrendered in rapid succession.   I guess that is somewhat offset by partitions being independently managed, so the handler calls can interleave between partitions as well, but does make this a bit harder to follow.

- Follow the "your code needs to manage its own exceptions and we make no guarantees on behavior" model and await the handler.  We'd still ignore the error, but we'd be opening the door to a performance hit as the user-code may be heavy and it may not be clear that this blocks the processor instance from restarting processing or finishing surrender so that another processor can claim.  On the upside, it makes the start/stop handlers more predictable because they won't interleave. 

- Await the task and surface through the event handler; this goes against the grain a small bit in that we're surfacing user-code issues from non-essential calls and also opens that performance issue.  It does makes the start/stop handlers more predictable because they won't interleave. 

I admit to being torn between the first and second, but don't have a strong opposition to the thir.   I'm leaning towards keeping the first approach in the spirit of trying to prioritize performance, but if you give me the stink eye long enough you can probably change my mind.

I'd be interested in your thoughts and rationale.

Reply:
> I'd be interested in your thoughts and rationale.

Of course!  

Personally I would prefer the awaiting model.  If I was concerned about about the throughput hit - I could always call `Task.Run(() => ....)` myself in the handler and immediately return `Task.CompletedTask`.  In that world, the consumer gets a simpler mental model to start with but they have an escape hatch if we made the wrong call with respect to performance.

For me the more deterministic model is appealing.  I think it also more closely aligns with the model of the Track 1 SDK.  I personally find it easier to reason about.

In that world, I would say: "Calls to Start/Stop/ProcessEvents for a given partition never run concurrently.   Start/Stop/ProcessEvents should not throw exceptions, but if there are exceptions, we deliver them to your error handler in a fire and forget fashion.  The error handler is able to interleave with anything.  If your error handler faults, you get whatever your hosting environment decides to do with unobserved failing tasks".  This feels reasonable to me and customers would grok it pretty quickly.

I suppose that actually delivering the error to the error handler is a little concerning, because if the error handler called `StopProcessing` maybe it's possible for us to get in a weird deadlock/infinite loop?  If that was a concern, I think I'd still at least prefer an await and log to the event source, just to keep a simpler model about the state machine this thing is running under the hood.

N.B. When I started writing this comment I agreed with your position.  My thought was "well, at least an instances of TPartition are never shared between concurrently running Start/Stop calls, so that's nice."  But it feels like it possible to recover much of the benefit of this model for cases where this ends up causing a throughput hit, by just having the user's code do the `Task.Run`.

Reply:
You bring up a point that I had not thought about, which is that I believe #3 does come with the potential for an infinite loop scenario if the stop handler triggers the error handler which asks to stop and then triggers the error in the stop handler.... 

That said, between your perspective and having given this more thought last night, I think that I'm convinced that we should try swapping to approach #2, where we await and log, though that will prevent us from calling the sync version of `StopProcessing` in this handler - which was something that may make sense to do.   Though, as you mention, it could be achieved by doing a `Task.Run` in the user-code for the handler.

Reply:
I've gone down the path of option 2 and augmented the comments for the handler methods to advise against attempts to directly control the state of the processor from within them.